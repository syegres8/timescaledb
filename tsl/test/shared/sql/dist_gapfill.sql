-- This file and its contents are licensed under the Timescale License.
-- Please see the included NOTICE for copyright information and
-- LICENSE-TIMESCALE for a copy of the license.

\set ECHO errors

\set TEST_BASE_NAME dist_gapfill
\set TEST_METRICS_NAME gapfill_metrics
SELECT format('include/%s_query.sql', :'TEST_BASE_NAME') AS "TEST_QUERY_NAME",
    format('%s/results/%s_singlenode.out', :'TEST_OUTPUT_DIR', :'TEST_BASE_NAME') AS "TEST_SINGLENODE",
    format('%s/results/%s_partitionwise_off.out', :'TEST_OUTPUT_DIR', :'TEST_BASE_NAME') AS "TEST_PARTITIONWISE_OFF",
    format('%s/results/%s_partitionwise_on.out', :'TEST_OUTPUT_DIR', :'TEST_BASE_NAME') AS "TEST_PARTITIONWISE_ON",
    format('include/%s_query.sql', :'TEST_METRICS_NAME') AS "TEST_METRICS_QUERY_NAME",
    format('%s/results/%s_nohyper.out', :'TEST_OUTPUT_DIR', :'TEST_METRICS_NAME') AS "TEST_METRICS_NOHYPER",
    format('%s/results/%s_partitionwise_off.out', :'TEST_OUTPUT_DIR', :'TEST_METRICS_NAME') AS "TEST_METRICS_PARTITIONWISE_OFF" \gset

SELECT format('\! diff %s %s', :'TEST_SINGLENODE', :'TEST_PARTITIONWISE_OFF') AS "DIFF_CMD_PARTITIONWISE_OFF",
    format('\! diff %s %s', :'TEST_SINGLENODE', :'TEST_PARTITIONWISE_ON') AS "DIFF_CMD_PARTITIONWISE_ON",
    format('\! diff %s %s', :'TEST_METRICS_NOHYPER', :'TEST_METRICS_PARTITIONWISE_OFF') AS "DIFF_CMD_METRICS_PARTITIONWISE_OFF" \gset

-- Non-distributed hypertables
-- dist_gapfill_query
\set CONDITIONS conditions

\set ECHO all
\ir :TEST_QUERY_NAME
\set ECHO errors
\o :TEST_SINGLENODE
\ir :TEST_QUERY_NAME
\o

-- Run gapfill on a table as in gapfill.sql, where the result is verified

\set METRICS metrics_int

\o :TEST_METRICS_NOHYPER
\ir :TEST_METRICS_QUERY_NAME
\o

-- Distributed hypertables with three data nodes

-- dist_gapfill_query

\set CONDITIONS conditions_dist

SET enable_partitionwise_aggregate = 'off';
\o :TEST_PARTITIONWISE_OFF
\ir :TEST_QUERY_NAME
\o

SET enable_partitionwise_aggregate = 'on';
\o :TEST_PARTITIONWISE_ON
\ir :TEST_QUERY_NAME
\o

SET enable_partitionwise_aggregate = 'off';

-- gapfill_metrics_query
\set METRICS metrics_int_dist

\o :TEST_METRICS_PARTITIONWISE_OFF
\ir :TEST_METRICS_QUERY_NAME
\o

\set ECHO all

:DIFF_CMD_PARTITIONWISE_OFF
:DIFF_CMD_PARTITIONWISE_ON
:DIFF_CMD_METRICS_PARTITIONWISE_OFF

\set ECHO errors

-- Distributed hypertables with one data nodes

-- dist_gapfill_query
\set CONDITIONS conditions_dist1

SET enable_partitionwise_aggregate = 'off';
\o :TEST_PARTITIONWISE_OFF
\ir :TEST_QUERY_NAME
\o

SET enable_partitionwise_aggregate = 'on';
\o :TEST_PARTITIONWISE_ON
\ir :TEST_QUERY_NAME
\o

SET enable_partitionwise_aggregate = 'off';

-- gapfill_metrics_query
\set METRICS metrics_int_dist1

\o :TEST_METRICS_PARTITIONWISE_OFF
\ir :TEST_METRICS_QUERY_NAME
\o

\set ECHO all

:DIFF_CMD_PARTITIONWISE_OFF
:DIFF_CMD_PARTITIONWISE_ON
:DIFF_CMD_METRICS_PARTITIONWISE_OFF
