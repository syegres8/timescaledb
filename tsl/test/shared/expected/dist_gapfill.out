-- This file and its contents are licensed under the Timescale License.
-- Please see the included NOTICE for copyright information and
-- LICENSE-TIMESCALE for a copy of the license.
\set ECHO errors
\ir :TEST_QUERY_NAME
-- This file and its contents are licensed under the Timescale License.
-- Please see the included NOTICE for copyright information and
-- LICENSE-TIMESCALE for a copy of the license.
SELECT time_bucket_gapfill('3 hours', time, '2017-01-01 06:00', '2017-01-02 18:00'),
       first(value, time),
       avg(value)
FROM :CONDITIONS
GROUP BY 1;
     time_bucket_gapfill      | first |   avg   
------------------------------+-------+---------
 Sun Jan 01 04:00:00 2017 PST |   1.2 |     1.2
 Sun Jan 01 07:00:00 2017 PST |   7.3 |     5.8
 Sun Jan 01 10:00:00 2017 PST |       |        
 Sun Jan 01 13:00:00 2017 PST |       |        
 Sun Jan 01 16:00:00 2017 PST |       |        
 Sun Jan 01 19:00:00 2017 PST |       |        
 Sun Jan 01 22:00:00 2017 PST |       |        
 Mon Jan 02 01:00:00 2017 PST |       |        
 Mon Jan 02 04:00:00 2017 PST |       |        
 Mon Jan 02 07:00:00 2017 PST |  0.23 |    0.23
 Mon Jan 02 10:00:00 2017 PST |       |        
 Mon Jan 02 13:00:00 2017 PST |       |        
 Mon Jan 02 16:00:00 2017 PST |       |        
 Sun Jul 01 05:00:00 2018 PDT |   3.1 |     3.1
 Sun Jul 01 08:00:00 2018 PDT |    64 | 5183.56
 Mon Jul 02 08:00:00 2018 PDT |     0 |       0
(16 rows)

SELECT time_bucket_gapfill('3 hours', time, '2017-01-01 06:00', '2017-01-01 18:00'),
       device,
       first(value, time),
       avg(value)
FROM :CONDITIONS
GROUP BY 1,2;
     time_bucket_gapfill      | device |  first   |   avg    
------------------------------+--------+----------+----------
 Sun Jan 01 04:00:00 2017 PST |      1 |      1.2 |      1.2
 Sun Jan 01 07:00:00 2017 PST |      1 |      7.3 |      7.3
 Sun Jan 01 10:00:00 2017 PST |      1 |          |         
 Sun Jan 01 13:00:00 2017 PST |      1 |          |         
 Sun Jan 01 16:00:00 2017 PST |      1 |          |         
 Sun Jan 01 04:00:00 2017 PST |      2 |          |         
 Sun Jan 01 07:00:00 2017 PST |      2 |          |         
 Sun Jan 01 10:00:00 2017 PST |      2 |          |         
 Sun Jan 01 13:00:00 2017 PST |      2 |          |         
 Sun Jan 01 16:00:00 2017 PST |      2 |          |         
 Mon Jan 02 07:00:00 2017 PST |      2 |     0.23 |     0.23
 Sun Jan 01 04:00:00 2017 PST |      3 |          |         
 Sun Jan 01 07:00:00 2017 PST |      3 |      4.3 |      4.3
 Sun Jan 01 10:00:00 2017 PST |      3 |          |         
 Sun Jan 01 13:00:00 2017 PST |      3 |          |         
 Sun Jan 01 16:00:00 2017 PST |      3 |          |         
 Sun Jan 01 04:00:00 2017 PST |     13 |          |         
 Sun Jan 01 07:00:00 2017 PST |     13 |          |         
 Sun Jan 01 10:00:00 2017 PST |     13 |          |         
 Sun Jan 01 13:00:00 2017 PST |     13 |          |         
 Sun Jan 01 16:00:00 2017 PST |     13 |          |         
 Sun Jul 01 05:00:00 2018 PDT |     13 |      3.1 |      3.1
 Sun Jan 01 04:00:00 2017 PST |     29 |          |         
 Sun Jan 01 07:00:00 2017 PST |     29 |          |         
 Sun Jan 01 10:00:00 2017 PST |     29 |          |         
 Sun Jan 01 13:00:00 2017 PST |     29 |          |         
 Sun Jan 01 16:00:00 2017 PST |     29 |          |         
 Sun Jul 01 08:00:00 2018 PDT |     29 |       64 |       64
 Sun Jan 01 04:00:00 2017 PST |     87 |          |         
 Sun Jan 01 07:00:00 2017 PST |     87 |          |         
 Sun Jan 01 10:00:00 2017 PST |     87 |          |         
 Sun Jan 01 13:00:00 2017 PST |     87 |          |         
 Sun Jan 01 16:00:00 2017 PST |     87 |          |         
 Mon Jul 02 08:00:00 2018 PDT |     87 |        0 |        0
 Sun Jan 01 04:00:00 2017 PST |     90 |          |         
 Sun Jan 01 07:00:00 2017 PST |     90 |          |         
 Sun Jan 01 10:00:00 2017 PST |     90 |          |         
 Sun Jan 01 13:00:00 2017 PST |     90 |          |         
 Sun Jan 01 16:00:00 2017 PST |     90 |          |         
 Sun Jul 01 08:00:00 2018 PDT |     90 | 10303.12 | 10303.12
(40 rows)

SELECT time_bucket_gapfill('3 hours', time, '2017-01-01 06:00', '2017-01-01 18:00'),
       device,
       first(value, time),
       avg(value)
FROM :CONDITIONS
GROUP BY 2,1;
     time_bucket_gapfill      | device |  first   |   avg    
------------------------------+--------+----------+----------
 Sun Jan 01 04:00:00 2017 PST |      1 |      1.2 |      1.2
 Sun Jan 01 07:00:00 2017 PST |      1 |      7.3 |      7.3
 Sun Jan 01 10:00:00 2017 PST |      1 |          |         
 Sun Jan 01 13:00:00 2017 PST |      1 |          |         
 Sun Jan 01 16:00:00 2017 PST |      1 |          |         
 Sun Jan 01 04:00:00 2017 PST |      2 |          |         
 Sun Jan 01 07:00:00 2017 PST |      2 |          |         
 Sun Jan 01 10:00:00 2017 PST |      2 |          |         
 Sun Jan 01 13:00:00 2017 PST |      2 |          |         
 Sun Jan 01 16:00:00 2017 PST |      2 |          |         
 Mon Jan 02 07:00:00 2017 PST |      2 |     0.23 |     0.23
 Sun Jan 01 04:00:00 2017 PST |      3 |          |         
 Sun Jan 01 07:00:00 2017 PST |      3 |      4.3 |      4.3
 Sun Jan 01 10:00:00 2017 PST |      3 |          |         
 Sun Jan 01 13:00:00 2017 PST |      3 |          |         
 Sun Jan 01 16:00:00 2017 PST |      3 |          |         
 Sun Jan 01 04:00:00 2017 PST |     13 |          |         
 Sun Jan 01 07:00:00 2017 PST |     13 |          |         
 Sun Jan 01 10:00:00 2017 PST |     13 |          |         
 Sun Jan 01 13:00:00 2017 PST |     13 |          |         
 Sun Jan 01 16:00:00 2017 PST |     13 |          |         
 Sun Jul 01 05:00:00 2018 PDT |     13 |      3.1 |      3.1
 Sun Jan 01 04:00:00 2017 PST |     29 |          |         
 Sun Jan 01 07:00:00 2017 PST |     29 |          |         
 Sun Jan 01 10:00:00 2017 PST |     29 |          |         
 Sun Jan 01 13:00:00 2017 PST |     29 |          |         
 Sun Jan 01 16:00:00 2017 PST |     29 |          |         
 Sun Jul 01 08:00:00 2018 PDT |     29 |       64 |       64
 Sun Jan 01 04:00:00 2017 PST |     87 |          |         
 Sun Jan 01 07:00:00 2017 PST |     87 |          |         
 Sun Jan 01 10:00:00 2017 PST |     87 |          |         
 Sun Jan 01 13:00:00 2017 PST |     87 |          |         
 Sun Jan 01 16:00:00 2017 PST |     87 |          |         
 Mon Jul 02 08:00:00 2018 PDT |     87 |        0 |        0
 Sun Jan 01 04:00:00 2017 PST |     90 |          |         
 Sun Jan 01 07:00:00 2017 PST |     90 |          |         
 Sun Jan 01 10:00:00 2017 PST |     90 |          |         
 Sun Jan 01 13:00:00 2017 PST |     90 |          |         
 Sun Jan 01 16:00:00 2017 PST |     90 |          |         
 Sun Jul 01 08:00:00 2018 PDT |     90 | 10303.12 | 10303.12
(40 rows)

SELECT
  time_bucket_gapfill('3 hours', time, '2017-01-01 06:00', '2017-01-01 18:00'),
  lag(min(time)) OVER ()
FROM :CONDITIONS
GROUP BY 1;
     time_bucket_gapfill      |             lag              
------------------------------+------------------------------
 Sun Jan 01 04:00:00 2017 PST | 
 Sun Jan 01 07:00:00 2017 PST | Sun Jan 01 06:01:00 2017 PST
 Sun Jan 01 10:00:00 2017 PST | Sun Jan 01 08:01:00 2017 PST
 Sun Jan 01 13:00:00 2017 PST | 
 Sun Jan 01 16:00:00 2017 PST | 
 Mon Jan 02 07:00:00 2017 PST | 
 Sun Jul 01 05:00:00 2018 PDT | Mon Jan 02 08:01:00 2017 PST
 Sun Jul 01 08:00:00 2018 PDT | Sun Jul 01 06:01:00 2018 PDT
 Mon Jul 02 08:00:00 2018 PDT | Sun Jul 01 08:01:00 2018 PDT
(9 rows)

\set ECHO errors
:DIFF_CMD_PARTITIONWISE_OFF
:DIFF_CMD_PARTITIONWISE_ON
:DIFF_CMD_METRICS_PARTITIONWISE_OFF
\set ECHO errors
:DIFF_CMD_PARTITIONWISE_OFF
:DIFF_CMD_PARTITIONWISE_ON
:DIFF_CMD_METRICS_PARTITIONWISE_OFF
